OpenRA.Mods.Common/Traits/Player/ParallelProductionQueue.cs

```diff
@@ -0,0 +1,77 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ParallelProductionQueueInfo : ProductionQueueInfo
+	{
+		public override object Create(ActorInitializer init) { return new ParallelProductionQueue(init, init.Self.Owner.PlayerActor, this); }
+	}
+
+	public class ParallelProductionQueue : ProductionQueue
+	{
+		public ParallelProductionQueue(ActorInitializer init, Actor playerActor, ParallelProductionQueueInfo info) : base(init, playerActor, info)
+		{
+		}
+
+		protected override void TickInner(Actor self, bool allProductionPaused)
+		{
+			if (!allProductionPaused)
+			{
+				for (var i = 0; i < Queue.Count;)
+				{
+					var first = Queue[0];
+
+					var before = first.RemainingTime;
+
+					if (!first.Paused)
+						first.Tick(playerResources);
+
+					foreach (var item in Queue.FindAll(a => a.Item == first.Item))
+					{
+						i++;
+						Queue.Remove(item);
+						Queue.Add(item);
+					}
+
+					if (first.RemainingTime != before)
+						return;
+				}
+			}
+		}
+
+		public override bool IsProducing(ProductionItem item)
+		{
+			return Queue.Contains(item);
+		}
+
+		public override int RemainingTimeActual(ProductionItem item)
+		{
+			var parallelBuilds = 0;
+			var added = new List<string>();
+
+			foreach (var i in Queue)
+			{
+				if (added.Contains(i.Item))
+					continue;
+
+				added.Add(i.Item);
```

Comment:
It should be possible to simplify this by using a `HashSet` instead of a list and then doing `if (!added.Add(i.Item))` (which returns false if the item already existed)